### DISCUSSION TITLE
Umbraco Tracking References feature

### DISCUSSION THREAD

**Discussion by patrickdemooij9** [leekelleher/umbraco-contentment#216] [Discussion]
With the 9.4 release Umbraco now uses a lot more logic related to references.
I don't think Contentment currently does anything with references right? I couldn't find anything about it in the code. I think there might be two places that should have references?

- The content blocks. Might be as easy as implementing this change (https://github.com/PerplexDigital/Perplex.ContentBlocks/pull/54/files) that I did for the Perplex ContentBlocks.
- Data List. I think this is probably the most interesting one. For one of our projects, we actually use the Data list to grab content items from various locations and then show them in the data list. So, ideally you would want to create a reference to those chosen items.

Let me know what you think! I can probably start with a PR for the content blocks if you want this functionality. The Data List might take some more planning as you can't be sure what sort of item is in there (probably some sort of interface to create your own relations?)

**Comment by patrickdemooij9** [comment]
And perhaps the `Render Macro`, but not quite sure what it does and if it is possible to have a reference to a macro

**Comment by leekelleher** [comment]
Thanks @patrickdemooij9, good suggestion. 👍 

Implementing `IDataValueReference` has been on [my ideas/to-do list since v1.0-beta](https://github.com/leekelleher/umbraco-contentment/blob/1.0.0-beta/.github/IDEAS.md?plain=1#L20). 😄 

As you say, Contentment's [Content Blocks](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/content-blocks.md) editor is a prime candidate for this. 👍 Similar with the [Render Macro](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/render-macro.md) editor.

[Notes](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/notes.md) and [Editor Notes](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/editor-notes.md) both use the RTE, so could potentially reference a content link, a media item or a macro. Would need to see how Umbraco core handles those. 🤔 

[Data List](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/data-list.md), would be dependent on the data-source configuration, e.g. "[Umbraco Content](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/data-sources/data-source--umbraco-content.md)" is the obvious one, but there are a bunch of other Umbraco-based data-sources, so it would most likely be something that is implemented at the `IDataListSource` level - like you say, as an optional interface _(along the lines of the `IDataListSourceValueConverter` interface)_, named something like `IDataListSourceValueReference`. 🤔 

Also, [Text Input](https://github.com/leekelleher/umbraco-contentment/blob/develop/docs/editors/text-input.md) editor has an option to use data-sources too.

---

I think this brain dump is probably why I kept putting it off. 😆
Still worth getting started with the simpler editors. 👍
